### PR TITLE
suite-state command forward compat.

### PR DESCRIPTION
**Pull request by hjoliver**
Address #5236 
Sibling of #5237 

**Check List**

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Applied any dependency changes to both `setup.cfg` and `conda-environment.yml`.
- [x] Tests are included (or explain why tests are not needed).
- [x] `CHANGES.md` entry included if this is a change that can affect users
- [x] [Cylc-Doc](https://github.com/cylc/cylc-doc) pull request opened if required at cylc/cylc-doc/pull/XXXX.
- [x] If this is a bug fix, PR should be raised against the relevant `?.?.x` branch.

### CODE REVIEW

**Comment by hjoliver** [comment]
Not sure if we really need this one (it'd be nice to avoid yet another final Cylc 7 release). 

#5237 is the critical one, for Cylc 8 research users triggering tasks off of Cylc 7 operational suites that haven't been migrated yet.

**Comment by dwsutherland** [comment]
For operations, this is important for migration, otherwise we'll be forced to upgrade downstream first.. Which we could do I suppose...

**Comment by hjoliver** [comment]
Can you give this a quick review and test @dwsutherland ? 

(Since we need it soon!)

**Comment by hjoliver** [comment]
(New tests are not needed at this point, for Cylc 7 IMO)

**Review by MetRonnie — CHANGES_REQUESTED** [review]
Missing the xtrigger part here?
https://github.com/cylc/cylc-flow/blob/e60e8f7a6d77256af0e312237c6c8f7bf4fd38f8/lib/cylc/xtriggers/suite_state.py#L49

**Comment by hjoliver** [comment]
Good spot @MetRonnie, thanks. 

Turns out I hadn't quite finished this off - another change was need to handled the change in task_outputs table format too.  (Well, 11 months is a long time...)

**Comment by hjoliver** [comment]
Testing:

Run with Cylc 8:
```ini
[scheduler]
    cycle point format = %Y
[scheduling]
    initial cycle point = 2010
    final cycle point = 2012
   [[graph]]
      P1Y = "foo"
[runtime]
   [[foo]]
      script = "cylc message x"
      [[[outputs]]]
          x = x
```
Now with Cylc 7, both of these work:
```console
cylc-7 $ cylc suite-state c8/runN --task=foo --point=2011 --message=x
polling for 'output: x': satisfied

cylc-7 $ cylc suite-state c8/runN --task=foo --point=2011 --status=succeeded
polling for 'succeeded': satisfied
```

And to test xtriggers polling the above Cylc 8 workflow:
```ini
# suite.rc
[cylc]
  cycle point format = %Y
[scheduling]
    initial cycle point = 2010
    final cycle point = 2012
    [[xtriggers]]
         upstream = suite_state(suite=rtx/c8/runN, task=foo, point=%(point)s, \
            message=x):PT10S
   [[dependencies]]
        [[[P1Y]]]
           graph = "@upstream => foo"
[runtime]
    [[foo]]
```

**Comment by hjoliver** [comment]
Should be good to go now.

File: lib/cylc/dbstatecheck.py [hunk 1]
```diff
@@ -123,11 +131,12 @@ def task_state_met(self, task, cycle, status=None, message=None):
...
-                for value in json.loads(outputs_str).values():
-                    if message == value:
-                        return True
-            return False
+            # This works for Cylc 7 and Cylc 8 outputs table format:
+            return any(
+                message == value
+                for outputs_str, in res
+                for value in json.loads(outputs_str)
+            )
```
Comment (MetRonnie) [inline lib/cylc/dbstatecheck.py:139]: Are you saying removing the `.values()` makes it work for Cylc 8 DBs? Surely that breaks Cylc 7?

Comment (MetRonnie) [reply lib/cylc/dbstatecheck.py:139]: Or was that just a mistake?
```suggestion
            # This works for Cylc 7 and Cylc 8 outputs table format:
            return any(
                message == value
                for outputs_str, in res
                for value in json.loads(outputs_str).values()
            )
```

Comment (hjoliver) [reply lib/cylc/dbstatecheck.py:139]: Damn it. I owe you a chocolate fish 🐟 

It wasn't a mistake - I took it from master, and I tested it. 
However, I tested it with our classic shortcut `x = x` outputs - i.e. with message and label the same 🤦 

So this is a bug on master too. Patch coming...

**Comment by hjoliver** [comment]
@MetRonnie - now it's done!!

File: lib/cylc/dbstatecheck.py (hunk 1 above)
Comment (MetRonnie) [reply lib/cylc/dbstatecheck.py]: > So this is a bug on master too.

You mean, only a bug on master/8.x? (Though nearly introduced to Cylc 7 in this PR)

**Review by MetRonnie — APPROVED** [review]
Tested

**Pull request merged**